File: C++/singleNumber.cpp
/**
 * @File    : singleNumber.cpp
 * @Brief   : 只出现一次的数字
 * @Link    : https://leetcode-cn.com/problems/single-number/
 * @Author  : Wei Li
 * @Date    : 2021-07-21
*/

#include <iostream>
#include <vector>

class Solution
{
public:
    int singleNumber(std::vector<int> &nums)
    {
        int ans = 0;
        for (const int &num : nums)
        {
            ans ^= num;
        }
        return ans;
    }
};


// -------------------------------
int main(int argc, char **argv)
{
    // std::vector<int> nums{2, 2, 1};
    std::vector<int> nums{4, 1, 2, 1, 2};

    Solution solution;
    auto single_number = solution.singleNumber(nums);
    std::cout << "The solution of this problem is " << single_number << std::endl;

    return 0;
}

File: Python/singleNumber.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: singleNumber.py
@Function: 只出现一次的数字
@Link: https://leetcode-cn.com/problems/single-number/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-21
"""

from typing import List
from functools import reduce


class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        return reduce(lambda x, y: x ^ y, nums)


# -------------------------
if __name__ == "__main__":
    nums = [2, 2, 1]
    # nums = [4, 1, 2, 1, 2]

    solution = Solution()
    single_number = solution.singleNumber(nums)
    print(f"The solution of this problem is {single_number}")
